galleries/examples/specialty_plots/ishikawa_diagram.py

```diff
@@ -108,82 +109,78 @@ def causes(data: list, cause_x: float, cause_y: float,
 
 def draw_body(data: dict):
     """
-    Place each section in its correct place by changing
+    Place each problem section in its correct place by changing
     the coordinates on each loop.
 
     Parameters
     ----------
     data : dict
-        The input data (can be list or tuple). ValueError is
-        raised if more than six arguments are passed.
+        The input data (can be a dict of lists or tuples). ValueError
+        is raised if more than six arguments are passed.
 
     Returns
     -------
     None.
 
     """
-    second_sections = []
-    third_sections = []
-    # Resize diagram to automatically scale in response to the number
-    # of problems in the input data.
-    if len(data) == 1 or len(data) == 2:
-        spine_length = (-2.1, 2)
-        head_pos = (2, 0)
-        tail_pos = ((-2.8, 0.8), (-2.8, -0.8), (-2.0, -0.01))
-        first_section = [1.6, 0.8]
-    elif len(data) == 3 or len(data) == 4:
-        spine_length = (-3.1, 3)
-        head_pos = (3, 0)
-        tail_pos = ((-3.8, 0.8), (-3.8, -0.8), (-3.0, -0.01))
-        first_section = [2.6, 1.8]
-        second_sections = [-0.4, -1.2]
-    else:  # len(data) == 5 or 6
-        spine_length = (-4.1, 4)
-        head_pos = (4, 0)
-        tail_pos = ((-4.8, 0.8), (-4.8, -0.8), (-4.0, -0.01))
-        first_section = [3.5, 2.7]
-        second_sections = [1, 0.2]
-        third_sections = [-1.5, -2.3]
-
-    # Change the coordinates of the annotations on each loop.
+    # Set the length of the spine according to the number of 'problem' categories.
+    length = (math.ceil(len(data) / 2)) - 1
+
+    spine_length = [-2.1 - length, 2 + length]
+    head_pos = [2 + length, 0]
+    tail_pos = [[-2.8 - length, 0.8], [-2.8 - length, -0.8], [-2.0 - length, -0.01]]
```

Comment:
```suggestion
    tail_pos = [[-2.8 - length, 0.8], [-2.8 - length, -0.8], [-2.0 - length, -0.01]]
    draw_spine(spine_length, head_pos, tail_pos)
```

Let's move drawing the spine up here, this is where all relevant input for the function is calculated. Also, it's the order , you would draw such a diagram by hand: first the spine, and then the problem categories.

Reply:
It may even be a good idea to pull all the positioning details in to the function, so that you do 
```
draw_spine(-2 - length, 2 + length)
```
which gives the length of the line.

Note that the left of `spine_length` is -0.1 compared to that, which I assume is so that the line reaches into the tail triangle. Otherwise the transition may not look smooth. But that should be an implementation detail, i.e. (or so - untested)

```
def draw_spine(xmin, xmax):
    ...
    # draw main spine
    ax.plot([xmin-0.1, xmax], [0, 0], color='tab:blue', linewidth=2)
    # draw fish head
    ax.text(xmax + 0.1, - 0.05, 'PROBLEM', fontsize=10,
            weight='bold', color='white')
    semicircle = Wedge((xmax, 0), 1, 270, 90, fc='tab:blue')
    ax.add_patch(semicircle)
    # draw fishtail
    tail_pos = [[xmin - 0.8, 0.8], [xmin - 0.8, -0.8], [xmin, -0.01]]
    triangle = Polygon(tail, fc='tab:blue')
    ax.add_patch(triangle)
```